README.md

# DYCornerBanner

DYCornerBanner written in **Swift 3** allows you to quickly create corner labels on top of any kind of view, e.g. UIImageView. The rotation angle of the corner label is determined by the width and height of the banner view. 


## Features

* set the corner position to bottom right, top right, top left or bottom left.
* supports banners in any rotation angle and distance from the corner - the rotation angle and the distance from the corner are determined by the size and height-width proportion of the DYCornerBanner view's frame. 
* override the default setting for the banner label width. 
* override the default setting for the banner text font.

## Installation


Install DYCornerBanner through Cocoapods or simply clone the repository and drop the DYCornerBanner.swift file into your project. Enter the following information into your Podfile (see current version in header):

```Ruby
platform :ios, '9.0'

use_frameworks!

target '[Your app project title]' do
pod 'DYCornerBanner', '~> 1.0'
end

```
Make sure to import DYCornerBanner into your View Controller subclass:

```Swift
	import DYCornerBanner
```

## Usage

See the following examples.

![DYCornerBanner examples](https://github.com/DominikButz/DYCornerBanner/blob/master/gitResources/cornerBannerExamples.png "Corner banner examples")

### Code example: Creating a DYCornerBanner from Interface Builder

```Swift
 @IBOutlet weak var cornerBannerBottomLeft: DYCornerBanner!  // simply drop a UIView into the VC from interface builder in storyboard  and set the UIView class to DYCornerBanner
 

 override func viewDidLoad() {
   super.viewDidLoad()
   
   self.cornerBannerBottomLeft.setup(text: "Sunshine Guaranteed!", textColor: 	UIColor.black, bannerColor: UIColor.yellow, position: .bottomLeft)
   
 }

```

### Code example: Creating a DYCornerBanner in code

Alternatively, you can create a corner banner instance in code like so: 
```Swift
   let frame = CGRect(x: self.imageView.frame.size.width -  90, y: 0, width: 90, height: 90)
   let bannerInCode = DYCornerBanner(frame: frame)
   bannerInCode.setup(text: "Free Cocktails!", textColor: UIColor.red, bannerColor: UIColor.white.withAlphaComponent(0.6), position: .topRight)
  self.imageView.addSubview(bannerInCode)
```


### Customising
You can override the settings for bannerWidth and textFont as shown below. Make sure to set these properties **before calling setup**, otherwise the defaults will be set instead. 

```Swift

  self.cornerBannerTopLeft.textFont = UIFont(name: "ChalkboardSE-Bold" , size: 15.0)!
  self.cornerBannerTopLeft.bannerWidth = 17.0
self.cornerBannerTopLeft.setup(text: "Just Added!", textColor: UIColor.white, bannerColor: UIColor.grassGreen(), position: .topLeft)

```

## License

The MIT License (MIT)

Copyright (c) 2017 Dominik Butz

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG.md
# Change log

## [Version 1.0](https://github.com/DominikButz/DYAlertController/releases/tag/1.0)
Released on 2017-04-12.
First release!




DYCornerBanner/DYCornerBanner.swift
//
//  DYCornerBanner.swift
//  Flat Chat
//
//  Created by Dominik Butz on 10.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import UIKit

public class DYCornerBanner: UIView {

    var bannerLabel:UILabel = UILabel()
    
    var bannerPosition: BannerPosition = .bottomRight
   
    
    fileprivate let configuration = Configuration()
    
    open var textFont: UIFont {
        get {
            return self.configuration.textFont
        }
        set(font) {
            self.configuration.textFont = font
        }
    }
    
    open var bannerWidth: CGFloat {
        get {
            return self.configuration.bannerWidth
        }
        set(width) {
            self.configuration.bannerWidth = width
        }
    }

    

    
    // This method gets called if you create the view in the Interface Builder
    required public init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
        print("init decoder")
            //self.checkSquareShape()
        
  
    }
    
    // This method gets called if you create the view in code
  public  override init(frame: CGRect){
        super.init(frame: frame)
    print("init in code with frame")
         //  self.checkSquareShape()

    }
    
    /// setup - Has to be called on corner banner instance after creating the instance either in code or in the storyboard
    ///
    /// - Parameters:
    ///   - text: Banner label text.
    ///   - textColor: text color.
    ///   - bannerColor: fill colour of the banner label.
    ///   - position: .bottomRight, .topRight, .topLeft or .bottomLeft
    public func setup(text: String, textColor: UIColor, bannerColor: UIColor, position:BannerPosition ) {
        print("setup called")
     
        // self
        self.backgroundColor = UIColor.clear
        self.isUserInteractionEnabled = false
        self.clipsToBounds = true

          self.bannerPosition = position

        // dummy label for size calculation
        let sizeAdjustmentLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.frame.width, height: self.frame.height / 4.0))
        sizeAdjustmentLabel.text = text
        sizeAdjustmentLabel.font = UIFont.systemFont(ofSize: self.bannerWidth)
        sizeAdjustmentLabel.sizeToFit()
//
        // banner size
    //    let labelWidth =  sqrt( pow(self.frame.width, 2) self.frame.width * self.frame.width  + self.frame.height * self.frame.height)
        let labelWidth =  sqrt( pow(self.frame.width, 2) +  pow(self.frame.height, 2))
        
        
        print("label width \(labelWidth)" )
        let labelHeight = sizeAdjustmentLabel.frame.size.height
            print("label height \(labelHeight)" )
        let labelFrame = CGRect(x: 0, y: 0, width: labelWidth, height: labelHeight)
        
        self.bannerLabel = UILabel(frame: labelFrame)
        print("\(self.bannerLabel.frame.width), height \(self.bannerLabel.frame.size.height) ")
//        
//        // banner config
        self.bannerLabel.text = text
        self.bannerLabel.textAlignment = .center
        self.bannerLabel.baselineAdjustment = .alignCenters
        self.bannerLabel.font = self.textFont
        self.bannerLabel.textColor = textColor
        self.bannerLabel.numberOfLines = 1
        self.bannerLabel.lineBreakMode = .byTruncatingTail
        self.bannerLabel.backgroundColor = bannerColor
    
        // banner positioning
        var xAdjustment: CGFloat
        var yAdjustment: CGFloat
        let halfLabelHeight: CGFloat  = labelHeight / 2.0
        
        switch self.bannerPosition  {
        case .bottomRight:
            xAdjustment = halfLabelHeight
            yAdjustment = halfLabelHeight
        case .topRight:
            xAdjustment = halfLabelHeight
            yAdjustment = -halfLabelHeight
        case .bottomLeft:
            xAdjustment = -halfLabelHeight
            yAdjustment = halfLabelHeight
        case .topLeft:
            xAdjustment = -halfLabelHeight
            yAdjustment = -halfLabelHeight
            
        }
        
        self.bannerLabel.center = CGPoint(x: self.frame.width / 2.0 + xAdjustment, y: self.frame.height / 2.0 + yAdjustment)
   
        print("label \(self.bannerLabel)")
        self.addSubview(self.bannerLabel)
        
        
       
    }
    
    override public func layoutSubviews() {
             super.layoutSubviews()
        print("layout subviews called")

        let sinRotationAngle = self.frame.size.height / self.bannerLabel.frame.size.width
        let arcRotation = asin(sinRotationAngle)
        print("arc rotation \(arcRotation)")
        //
        let rotationAngleFactor = { () -> CGFloat in
            switch self.bannerPosition {
            case .bottomRight, .topLeft:
                return -1.0
            case .bottomLeft, .topRight:
                return 1.0
            }
        }()

        let rotation = CGAffineTransform.init(rotationAngle: CGFloat(rotationAngleFactor * arcRotation))
        
        self.bannerLabel.transform = rotation
        
   

    }
    


}



/// an enum for setting the corner banner position
///
/// - bottomRight:  surprise! bottomRight corner.
/// - topRight: topRight
/// - topLeft: yes, you guessed it: top left corner!
/// - bottomLeft: wow, this one, too!
public enum BannerPosition {
    case bottomRight, topRight, topLeft, bottomLeft
}



fileprivate class Configuration {
    
    var textFont: UIFont!
    var bannerWidth: CGFloat!
    
    init() {
       self.setupDefaults()
    }
    
    /// defaults for bannerWidth and text font. The properties in the corner label instance can be overridden.
    func setupDefaults() {
        
        self.bannerWidth = 12.0
        self.textFont = UIFont(name: "HelveticaNeue-Medium", size: 10.0)
        
    }
    
    
}


DYCornerBannerTests/DYCornerBannerTests.swift
//
//  DYCornerBannerTests.swift
//  DYCornerBannerTests
//
//  Created by Dominik Butz on 11.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import XCTest
import UIKit

@testable import DYCornerBanner

class DYCornerBannerTests: XCTestCase {
    
    var imageView:UIImageView?
    var cornerBannerBottomRight : DYCornerBanner?
    
    override func setUp() {
        super.setUp()
        
        self.imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 400, height: 200))
            
        self.cornerBannerBottomRight = DYCornerBanner(frame: CGRect(x: 240, y: 110, width: 160, height: 90))
                
    self.imageView?.image = UIImage(named:"IMG_0002.jpg")
        self.imageView?.addSubview(self.cornerBannerBottomRight!)
      
        
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
        
        self.imageView = nil
        self.cornerBannerBottomRight = nil
  
    }
    
    func testCornerBannerBottomRight() {
        
   
        
        self.cornerBannerBottomRight?.textFont = UIFont(name: "AmericanTypewriter-Bold", size: 11.0)!
        self.cornerBannerBottomRight?.bannerWidth = 17.0
        self.cornerBannerBottomRight?.setup(text: "Top Cruise Tour!", textColor: UIColor.white, bannerColor: UIColor.red, position: .bottomRight)
        
        XCTAssert(self.cornerBannerBottomRight?.bannerPosition == .bottomRight, "corner banner is wrong")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.font.familyName, "American Typewriter", "wrong font name!")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerWidth, 17.0, "Banner width value not correct")
           XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.text, "Top Cruise Tour!", "Banner label text incorrect!")
           XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.textColor, UIColor.white, "Banner label text color value not correct")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.backgroundColor, UIColor.red, "Banner label background color value not correct")

        
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


DYCornerBannerDemo/DYCornerBannerDemo/AppDelegate.swift
//
//  AppDelegate.swift
//  DYCornerBannerDemo
//
//  Created by Dominik Butz on 11.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



DYCornerBannerDemo/DYCornerBannerDemo/ViewController.swift
//
//  ViewController.swift
//  DYCornerBannerDemo
//
//  Created by Dominik Butz on 11.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import UIKit
import DYCornerBanner

class ViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    
    @IBOutlet weak var cornerBannerBottomRight: DYCornerBanner!
    
    @IBOutlet weak var cornerBannerTopLeft: DYCornerBanner!
    
    @IBOutlet weak var cornerBannerBottomLeft: DYCornerBanner!
    

    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        // with custom Font
        self.cornerBannerBottomRight.textFont = UIFont(name: "AmericanTypewriter-Bold", size: 11.0)!
        self.cornerBannerBottomRight.setup(text: "Top Cruise Tour!", textColor: UIColor.white, bannerColor: UIColor.red, position: .bottomRight)
        
        
        // with custom font and custom width
        self.cornerBannerTopLeft.textFont = UIFont(name: "ChalkboardSE-Bold" , size: 15.0)!
        self.cornerBannerTopLeft.bannerWidth = 17.0
        self.cornerBannerTopLeft.setup(text: "Just Added!", textColor: UIColor.white, bannerColor: UIColor.grassGreen(), position: .topLeft)

       
        
        self.cornerBannerBottomLeft.setup(text: "Sunshine Guaranteed!", textColor: UIColor.black, bannerColor: UIColor.yellow, position: .bottomLeft)
        
        
        // programatically created DY Corner Banner
        let frame = CGRect(x: self.imageView.frame.size.width -  90, y: 0, width: 90, height: 90)
        let bannerInCode = DYCornerBanner(frame: frame)
       bannerInCode.textFont = UIFont(name: "HelveticaNeue-Bold", size: 10.0)!
        bannerInCode.setup(text: "Free Cocktails!", textColor: UIColor.red, bannerColor: UIColor.white.withAlphaComponent(0.6), position: .topRight)
        self.imageView.addSubview(bannerInCode)
        

        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}


public extension UIColor {
    
    class  func veryDarkGray()->UIColor {
        
        return UIColor(red: 64/255, green: 64/255, blue: 65/255, alpha: 1.0)
    }
    
    class func veryLightGray()->UIColor {
        
        return UIColor(red: 239/255, green: 239/255, blue: 244/255, alpha: 1.0)
    }
    
    
    
    class func dirtyWhite()->UIColor {
        
        return UIColor(red: 250/255, green: 250/255, blue: 250/255, alpha: 1.0)
    }
    
    class func grassGreen()->UIColor {
        return UIColor(red: 39/255, green: 161/255, blue: 60/255, alpha: 1.0)
    }
    
    
    
    
}

